### ISSUE TITLE
Multiple profiles

### ISSUE THREAD

**Issue by tb102122** [aws-actions/configure-aws-credentials#112]
Is it possible to also setup multiple profiles which can be used?

**Comment by peterkline-es** [comment]
This would be extremely valuable for self-hosted runners where we have a single named user but it doesn't different things in different accounts

**Comment by ssannidhi-palo** [comment]
**Usecase:**
We are using terraform to setup our infrastructure in multiple aws accounts(one account for PROD, one account for non-prod envs). We maintain the state file of each env in S3 bucket of respective account. Prod env needs to access resources that are in a different aws account. Terraform provides a way to access remote state files present in s3 bucket of another aws account using "terraform_remote_state" with "**profile**".

data "terraform_remote_state" "env-GLOBAL" {
  backend = "s3"

  config = {
    profile = "demo"
    bucket  = "xxx"
    key     = "global/terraform.tfstate"
    region  = local.region
  }
}

To run the terraform scripts through Github Actions for such a use case, it would be very helpful if **aws-actions/configure-aws-credentials** provides a way to configure multiple profiles.

**Comment by Startouf** [comment]
I believe this fits the issue, we have a project with a specific AWS profile name used with serverless (helps developers segregate their AWS credentials across multipleprojects), it would be nice to allow specifying the profile name when storing the Access key and secret access key

Currently we resort to the following

```yaml
- name: Add profile credentials to ~/.aws/credentials
   env:
      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
   run: |
      aws configure set aws_access_key_id $AWS_ACCESS_KEY_ID --profile my-app-name
      aws configure set aws_secret_access_key $AWS_SECRET_ACCESS_KEY --profile my-app-name
```

**Comment by dennis-tra** [comment]
Untested, but wouldn't the following more concise version work as well?
```yaml
- name: Add profile credentials to ~/.aws/credentials 
   run: |
      aws configure set aws_access_key_id ${{ secrets.AWS_ACCESS_KEY_ID }} --profile my-app-name
      aws configure set aws_secret_access_key ${{ secrets.AWS_SECRET_ACCESS_KEY }} --profile my-app-name
```

**Comment by gastonsilva** [comment]
Extending @dennis-tra answer, you should also add:
`aws configure set aws_session_token $AWS_SESSION_TOKEN --profile my-app-name`
for this to work

**Comment by peterwoodworth** [comment]
I'm having trouble figuring out exactly how this would work with this action. When people want this action to support multiple profiles, how would you expect it to accomplish that? What is a sample input you could provide?

**Comment by pecigonzalo** [comment]
@peterwoodworth I think something like this would be ideal and simple:

```
    - name: Configure AWS credentials from Test account
      uses: aws-actions/configure-aws-credentials@v1
      with:
        role-to-assume: arn:aws:iam::111111111111:role/my-github-actions-role-test
        aws-region: us-east-1
        profile: foo
```

To be honest, I think overall supporting profile configs would be great.

We are currently emulating this with something like:
```
      - name: Configure AWS Profile
        run: |
          curl -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" "$ACTIONS_ID_TOKEN_REQUEST_URL&audience=sts.amazonaws.com" | jq -r '.value' > /tmp/.awscreds
          aws configure set region eu-central-1 --profile foo
          aws configure set role_arn "arn:aws:iam::123123123:role/foo" --profile foo
          aws configure set web_identity_token_file "/tmp/.awscreds" --profile foo
```

but this is fairly rudimentary

**Comment by pecigonzalo** [comment]
I've created a draft of this (not fully integrated) in https://github.com/aws-actions/configure-aws-credentials/pull/557

It creates a `exportCredentialsToConfig` that can add/modify profiles in `~/.aws/config` to configure different profiles.

If this is an approach you would be happy with, Im happy to finish the code and ready the PR.

**Comment by rafaelfuchs-mb** [comment]
Have you tested this solution?
I'm doing something similar here, but I'm still getting an error.

`│ Error: error configuring S3 Backend: no valid credential sources for S3 Backend found.`

workflow file:

`name: test-terraform-dns
run-name: ${{ github.actor }} is testing GitHub Actions
on: [push]
jobs:
  Terraform-run-all:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3
      - name: setup aws credentials
        run: |
          mkdir -p ~/.aws
          touch ~/.aws/credentials
          touch ~/.aws/config

          echo "
          [myprofile]
          aws_access_key_id = ${{ secrets.MY_AWS_ACCESS_KEY_ID }}
          aws_secret_access_key = ${{ secrets.MY_AWS_SECRET_ACCESS_KEY }}
          " > ~/.aws/credentials

          echo "
          [profile myprofile]
          region = us-west-2
          " > ~/.aws/config

      - name: check creds
        run: |
          cat ~/.aws/credentials
          cat ~/.aws/config
          aws configure list-profiles
      - name: Service - Terraform init all
        run: terraform init`

tf backend file:

`terraform {
  backend "s3" {
    bucket         = "terraform-s3-state-msysmc"
    key            = "path/to/my/state/terraform.tfstate"
    region         = "us-west-2"
    dynamodb_table = "terraform-dynamodb-locks"
    encrypt        = true
    profile        = "myprofile"
  }
}
`

**Comment by pecigonzalo** [comment]
@rafaelfuchs-mb are you asking about my solution or which one? If you are talking about what I linked in https://github.com/aws-actions/configure-aws-credentials/issues/112#issuecomment-1273757307, your workflow is not following my example, so it would not work.

**Comment by rafaelfuchs-mb** [comment]
You can ignore my comment... I was looking for a solution for multiple profiles with terraform, but it looks like something else is not working... I was able to get access to multiple profiles using aws cli, so my problem is actually in the terraform side. If that's ok, I can delete my previous comment.

**Comment by pecigonzalo** [comment]
@peterwoodworth let me know if the approach sounds good, so I can finalize the PR.

**Comment by peterwoodworth** [comment]
@pecigonzalo thanks for your patience, I'm hopping off for the day but I'll take a look tomorrow

**Comment by GeertWille** [comment]
@peterwoodworth @pecigonzalo any news regarding this issue and/or linked PR?

**Comment by nsandary** [comment]
I'm also interested in this functionality. We're building an action to synchronize RDS snapshots between multiple accounts, and it needs to access both accounts in the same workflow.

**Comment by pecigonzalo** [comment]
Im waiting on feedback from Peter

**Comment by GeertWille** [comment]
I created a PR to support this. Not sure or will be approved soon. In the meantime feel free to point to our branch and use it.

**Comment by pecigonzalo** [comment]
I mean, thanks for also taking a stab at it, glad to so a finalized solution. I would say tho that it's not polite to open a new PR @GeertWille given I had an open PR and was proposing a solution, and you even asked a question on my PR on how it works.

**Comment by GeertWille** [comment]
> I mean, thanks for also taking a stab at it, glad to so a finalized solution. I would say tho that it's a bit rude to open a new PR @GeertWille given I had an open PR and was proposing a solution, and you even asked a question on my PR on how it works.

Euh woops, definitely didn't ment to step on toes here. I can close mine again. Just needed a working solution. If your PR works with only adding a config instead of config and credentials I think yours is better. Just didn't wanted to wait on feedback due to lack of activity here and I need to have this working.

**Comment by GeertWille** [comment]
Still have an issue in my PR. Since the credentials are always exported it looks like precedence is given to the environment vars instead of the config files. So if you define 2 profiles. The last exported credentials are saved on the environment and used.

**Comment by woon-werd** [comment]
this works for me:

use aws plugin to login: 
```
    steps:
      - uses: actions/checkout@v3
      - name: Configure AWS credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          role-to-assume: arn:aws:iam::${{ env.ACCOUNT }}:role/${{ env.roleName }}
          aws-region: ${{ env.REGION }}
          role-session-name: ${{ env.roleName }}
```
setup profile with credentials from plugin - the plugin setup above automatically set credentials to env.*
```
      - name: Setup Profile
        run: |
          aws configure set region ${{ env.REGION }} --profile ${{ env.profileName }}
          aws configure set role_arn "arn:aws:iam::${{ env.ACCOUNT }}:role/${{ env.roleName }}" --profile ${{ env.profileName }}
          aws configure set role_session_name ${{ env.roleName }} --profile ${{ env.profileName }}
          aws configure set aws_access_key_id ${{ env.AWS_ACCESS_KEY_ID }} --profile ${{ env.profileName }}
          aws configure set aws_secret_access_key ${{ env.AWS_SECRET_ACCESS_KEY }} --profile ${{ env.profileName }}
          aws configure set aws_session_token ${{ env.AWS_SESSION_TOKEN }} --profile ${{ env.profileName }}

**Comment by stephendarling** [comment]
> this works for me:
> 
> use aws plugin to login:
> 
> ```
>     steps:
>       - uses: actions/checkout@v3
>       - name: Configure AWS credentials
>         uses: aws-actions/configure-aws-credentials@v1
>         with:
>           role-to-assume: arn:aws:iam::${{ env.ACCOUNT }}:role/${{ env.roleName }}
>           aws-region: ${{ env.REGION }}
>           role-session-name: ${{ env.roleName }}
> ```
> 
> setup profile with credentials from plugin - the plugin above automatically set credentials to env.*
> 
> ```
>       - name: Setup Profile
>         run: |
>           aws configure set region ${{ env.REGION }} --profile ${{ env.profileName }}
>           aws configure set aws_access_key_id ${{ env.AWS_ACCESS_KEY_ID }} --profile ${{ env.profileName }}
>           aws configure set aws_secret_access_key ${{ env.AWS_SECRET_ACCESS_KEY }} --profile ${{ env.profileName }}
>           aws configure set aws_session_token ${{ env.AWS_SESSION_TOKEN }} --profile ${{ env.profileName }}
> ```

@woon-werd, not sure if it's the same on `aws-actions/configure-aws-credentials@v1` (I'm using `aws-actions/configure-aws-credentials@v1-node16`) but the environment variable exported for region is `AWS_REGION`, not `REGION`. Works otherwise, thanks for this!

**Comment by mcblair** [comment]
I'm standing on the shoulders of giants with this, but here is something that I whipped up to meet my use case: https://github.com/marketplace/actions/configure-aws-profile

**Comment by woon-werd** [comment]
> > this works for me:
> > use aws plugin to login:
> > ```
> >     steps:
> >       - uses: actions/checkout@v3
> >       - name: Configure AWS credentials
> >         uses: aws-actions/configure-aws-credentials@v1
> >         with:
> >           role-to-assume: arn:aws:iam::${{ env.ACCOUNT }}:role/${{ env.roleName }}
> >           aws-region: ${{ env.REGION }}
> >           role-session-name: ${{ env.roleName }}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > setup profile with credentials from plugin - the plugin above automatically set credentials to env.*
> > ```
> >       - name: Setup Profile
> >         run: |
> >           aws configure set region ${{ env.REGION }} --profile ${{ env.profileName }}
> >           aws configure set aws_access_key_id ${{ env.AWS_ACCESS_KEY_ID }} --profile ${{ env.profileName }}
> >           aws configure set aws_secret_access_key ${{ env.AWS_SECRET_ACCESS_KEY }} --profile ${{ env.profileName }}
> >           aws configure set aws_session_token ${{ env.AWS_SESSION_TOKEN }} --profile ${{ env.profileName }}
> > ```
> 
> @woon-werd, not sure if it's the same on `aws-actions/configure-aws-credentials@v1` (I'm using `aws-actions/configure-aws-credentials@v1-node16`) but the environment variable exported for region is `AWS_REGION`, not `REGION`. Works otherwise, thanks for this!

ah yes. I hardcoded a few on them on my github actions. there are the useful things exported from the plugin:
env.AWS_ACCESS_KEY_ID 
env.AWS_SECRET_ACCESS_KEY
env.AWS_SESSION_TOKEN

Just spotted them on the log line whenever the plugin finish its works.

**Comment by kidpollo** [comment]
> I'm standing on the shoulders of giants with this, but here is something that I whipped up to meet my use case: https://github.com/marketplace/actions/configure-aws-profile

This solution WORKS. Thanks @mcblair I do still feel this should be part of this repo

**Comment by jonworek** [comment]
My use case was to create the profile which is being expected with the elastic beanstalk CLI which I'm using to deploy.

In the latest v2 version of the plugin, this is ultimately what I got to work:

```
      - name: Configure AWS credentials
        uses: aws-actions/configure-aws-credentials@v2
        with: 
          role-to-assume: arn:aws:iam::1233456788:role/GitHubActionsRole
          aws-region: us-east-2

      - name: Add profile credentials to ~/.aws/credentials 
        run: |
          aws configure set aws_access_key_id ${{ env.AWS_ACCESS_KEY_ID }} --profile eb-cli
          aws configure set aws_secret_access_key ${{ env.AWS_SECRET_ACCESS_KEY }} --profile eb-cli
          aws configure set aws_session_token ${{ env.AWS_SESSION_TOKEN }} --profile eb-cli
```

Thanks to all of the folks that commented above that steered me in this direction.

**Comment by Hronom** [comment]
Please, someone complete the work for this repo, from user side such API will be very needed and usable for terraform etc.
```
      - name: Configure AWS Credentials for profile-a
        uses: aws-actions/configure-aws-credentials@v2
        with:
          role-to-assume: ${{ inputs.profile_a_aws_arn }}
          aws-region: ${{ inputs.region }}
          profile: profile-a

      - name: Configure AWS Credentials for profile-b
        uses: aws-actions/configure-aws-credentials@v2
        with:
          role-to-assume: ${{ inputs.profile_b_aws_arn }}
          aws-region: ${{ inputs.region }}
          profile: profile-b

      - name: Configure AWS Credentials for profile-c
        uses: aws-actions/configure-aws-credentials@v2
        with:
          role-to-assume: ${{ inputs.profile_c_aws_arn }}
          aws-region: ${{ inputs.region }}
          profile: profile-c
```

3 years of discussions and still nothing for seems like basic functionality...

**Comment by ffMathy** [comment]
I need this too.

When we deploy to production, we'd like to also configure the staging environment credentials under the "staging" profile, so that our tests can then explicitly use that profile.

That way, it becomes easier to not accidentally run tests towards production.

**Comment by zilahir** [comment]
+1, this would be good!

**Comment by benjefferies** [comment]
> I'm standing on the shoulders of giants with this, but here is something that I whipped up to meet my use case: https://github.com/marketplace/actions/configure-aws-profile

Amazing stuff @mcblair. Needed to be done. Hopefully, someone can review the good work @GeertWille did https://github.com/aws-actions/configure-aws-credentials/pull/633

**Comment by ruckc** [comment]
We have a deployment that sprawls across multiple AWS accounts... it would be nice to be able to merge all of the tasks into a action.

**Comment by asychev** [comment]
Really, more than 3 years 🤯 ... maybe @tim-finnigan can help moving this forward?

**Comment by kokuyouwind** [comment]
Currently, I use a configuration file with `credential_source = Environment` as follows.
I personally recommend this method because it simplifies the workflow files by only requiring a fixed aws config to be placed in `.aws/config`.

`.github/workflows/aws_config/config`:
```
[profile a]
credential_source = Environment
role_arn = arn:aws:iam::xxxxx28:role/SwitchRoleA

[profile b]
credential_source = Environment
role_arn = arn:aws:iam::xxxxx02:role/SwitchRoleB
```

`.github/workflows/workflow.yml`:
```yml
    steps:
      - uses: aws-actions/configure-aws-credentials@v4
        with:
          role-to-assume: ${{ env.AWS_ROLE_ARN }}
      - name: Configure AWS Profiles
        run: |
          mkdir ~/.aws
          cp .github/workflows/aws_config/config ~/.aws/config
      - run: aws --profile=a sts get-caller-identity
      - run: aws --profile=b sts get-caller-identity
```

**Comment by Moulick** [comment]
[Building on the work of @mcblair](https://github.com/aws-actions/configure-aws-credentials/issues/112#issuecomment-1475388914), I have published an action that handles all of this very cleanly, it continues to use the official `aws-actions/configure-aws-credentials` under the hood + a couple of steps to see if all the profiles are correctly logged in correctly https://github.com/marketplace/actions/configure-multiple-aws-roles.

Also documented is the case on how to use with some composite docker steps that you might have later in the chain. Dependabot is enabled so any updates to the official action will be updated here asap.

**Comment by abseht** [comment]
A very much needed feature 🙏🏼 
💡 A use case for a least-privilege, multi-role, multi-account... Terraform access: ⬇️ 
Say we have a repository. It consists of multiple parent modules with simple configuration as below. Only the `profile` value is different and each relies on a dedicated role to manage AWS resources of a specific parent module:
```
terraform {
  required_version = ">= 1.3"

  backend "s3" {
    bucket                    = "terraform-state"
    key                          = "main.tfstate"
    region                    = "us-east-1"
    dynamodb_table = "terraform-state-lock"
    profile                    = "state-dev"
  }

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5"
    }
  }
}

provider "aws" {
  region = "us-east-1"
  profile = "dev"
...
}
```
Suggested look of the action:
```
steps:
  - uses: aws-actions/configure-aws-credentials@v4
    with:
      profiles:
        - name: "state-dev"
          role-to-assume: "your arn"
          aws-region: "us-west-1"
        - name: "dev"
          role-to-assume: "your other arn"
          aws-region: "us-east-1"
    ...other parameters...
```
As a result, with only one step, you have an environment which can seamlessly run commands for multiple Terraform parent modules.
🙏🏼 Thank you for your considerations!

**Comment by wonko** [comment]
As I needed this myself, and the ENV-vars are a serious problem in the GH actions, I created a very simple GH action which will do just that (take GH OIDC token, get a AWS session, store it to a named profile).

See https://github.com/krane-labs/gha-aws-oidc-profiles . You'd have to call it multiple times, each time with a role, a region, a sessionname and a profilename. Aftwards, the profiles can be used anywhere (terraform, but also cli with `aws xxx --profile yyy`. Might be useful to others.

**Comment by nmguse-bighealth** [comment]
Found this thread after suffering from the same limitation. I wanted an option that allowed us to use an aws config file and not rely on a third party action, and found the following to work (it assumes you setup your iam roles/identity provider in the same exact way as you would otherwise for aws-actions/configure-aws-credentials, since we already use that pattern here)

In the job:
```
permissions:
      # https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs
      contents: read  # This is required for actions/checkout
      id-token: write # Required for AWS Access

jobs:
  do_something:
    name: Do something
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Get AWS OIDC token
        run: |
            curl -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" "$ACTIONS_ID_TOKEN_REQUEST_URL&audience=sts.amazonaws.com" | jq -r '.value' > /tmp/awscreds

      - name: Setup AWS config
        run: mkdir ~/.aws && cp .github/aws_config ~/.aws/config

      - run: aws sts get-caller-identity
        env:
          AWS_PROFILE: <some profile name>
```

And in `.github/aws_config`
```
[profile default]
region = us-east-1
s3 = 
	max_concurrent_requests = 100

[profile <some profile name>]
web_identity_token_file = /tmp/awscreds
role_arn = <role arn>
```

**Comment by skokado** [comment]
Hi!

Based on this comment https://github.com/aws-actions/configure-aws-credentials/issues/112#issuecomment-1080050624
and refer from README example https://github.com/aws-actions/configure-aws-credentials?tab=readme-ov-file#assumerole-with-role-previously-assumed-by-action-in-same-workflow,

the following code worked for my case.

```yaml
# --- profile-a
- name: Configure AWS Credentials (profile-a)
  uses: aws-actions/configure-aws-credentials@v5.1.0
  id: creds-a
  with:
    aws-region: us-east-1
    role-to-assume: xxx
    output-credentials: true

- name: Set AWS profile (profile-a)
  env:
    profile: profile-a
  run: |
    aws --profile ${{ env.profile }} configure set aws_access_key_id     ${{ steps.creds-prod.outputs.aws-access-key-id }}
    aws --profile ${{ env.profile }} configure set aws_secret_access_key ${{ steps.creds-prod.outputs.aws-secret-access-key }}
    aws --profile ${{ env.profile }} configure set aws_session_token     ${{ steps.creds-prod.outputs.aws-session-token }}

# --- profile-b (default)
- name: Configure AWS Credentials (default)
  uses: aws-actions/configure-aws-credentials@v5.1.0
    aws-region: us-east-1
    role-to-assume: yyy

- name: debug
  run: |
    aws sts get-caller-identity
    aws sts get-caller-identity --profile profile-a
```

I hope this is helpful.

I strongly agree with this issue. I hope it will be easier to configure multiple profiles using the `profile` argument.
